test: use address_to_scriptpubkey instead of RPC call

ismaelsadeeq opened the pull request:
PR #27269 enables the function address_to_scriptpubkey() to decode all address types and return their corresponding scriptpubkeys. As a result, there is no longer any need to call getaddressinfo or validateaddress RPCs in order to retrieve an address scriptpubkey, as explained in the comments on this pull request (see https://github.com/bitcoin/bitcoin/pull/27269#pullrequestreview-1353681933 and https://github.com/bitcoin/bitcoin/pull/27269#issuecomment-1481016118).

Instead of using RPC calls, this update replaces the process of obtaining an address scriptPubkey with the address_to_scriptpubkey method, resulting in improved performance for functional tests.

ismaelsadeeq commented:
@josibake  @theStack @willcl-ark this is PR https://github.com/bitcoin/bitcoin/pull/27269 follow up, I will appreciate your reviews.
Thank you

josibake commented on test/functional/feature_nulldummy.py:
```diff
@@ -77,7 +78,8 @@ def run_test(self):
         self.ms_address = cms["address"]
-        ms_unlock_details = {"scriptPubKey": self.nodes[0].validateaddress(self.ms_address)["scriptPubKey"],
+        spk = address_to_scriptpubkey(self.ms_address).hex()
+        ms_unlock_details = {"scriptPubKey":spk,
```
I'm not sure adding `spk` as a variable really helps anything here and it does add an extra line to be reviewed. For a refactor, I'd suggest keeping the diff as small and clear as possible

```suggestion
        ms_unlock_details = {"scriptPubKey": address_to_scriptpubkey(self.ms_address).hex(),
```

ismaelsadeeq replied:
Thanks.

josibake reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/27349/commits/c73a0658c2c5c43d4ba2d8dbf2af71abe84b40e8

Nice to see a standard function for getting the `scriptPubKey` instead of a mix of RPC calls. Removing the RPC call _should_ make the tests run faster, but considering it wasn't used in many places, I doubt it makes a material difference. Left two non-blocking nits.

I also noticed the CI was failing, but when I rebased on master and ran the failing functional test locally, everything passed.

ismaelsadeeq commented on test/functional/test_framework/wallet.py:
```diff
@@ -101,7 +97,7 @@ def __init__(self, test_node, *, mode=MiniWalletMode.ADDRESS_OP_TRUE):
         elif mode == MiniWalletMode.ADDRESS_OP_TRUE:
             self._address, self._internal_key = create_deterministic_address_bcrt1_p2tr_op_true()
-            self._scriptPubKey = bytes.fromhex(self._test_node.validateaddress(self._address)['scriptPubKey'])
+            self._scriptPubKey = address_to_scriptpubkey(self._address)
```
Yes.

ismaelsadeeq commented:
> @ismaelsadeeq seems ready for review to me
Ready for review now.

ismaelsadeeq commented:
> ACK [c73a065](https://github.com/bitcoin/bitcoin/commit/c73a0658c2c5c43d4ba2d8dbf2af71abe84b40e8)
> 
> Nice to see a standard function for getting the `scriptPubKey` instead of a mix of RPC calls. Removing the RPC call _should_ make the tests run faster, but considering it wasn't used in many places, I doubt it makes a material difference. Left two non-blocking nits.
> 
> I also noticed the CI was failing, but when I rebased on master and ran the failing functional test locally, everything passed.

Thank you for the review Josi
I rebased to master recently but it failed again.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/27349/commits/01e0a1679fb1deeae4daf674730de2e1a270139e

thanks for taking the suggestions! CI failure seems unrelated from what I can tell :shrug:

josibake commented on test/functional/feature_nulldummy.py:
```diff
@@ -77,7 +78,7 @@ def run_test(self):
         wms = self.nodes[0].createmultisig(1, [self.pubkey], 'p2sh-segwit')
         self.ms_address = cms["address"]
-        ms_unlock_details = {"scriptPubKey": self.nodes[0].validateaddress(self.ms_address)["scriptPubKey"],
+        ms_unlock_details = {"scriptPubKey":address_to_scriptpubkey(self.ms_address).hex(),
```
nit: should be a space here

```suggestion
        ms_unlock_details = {"scriptPubKey": address_to_scriptpubkey(self.ms_address).hex(),
```

ismaelsadeeq replied:
Thanks

theStack commented on test/functional/test_framework/address.py:
```diff
@@ -21,11 +21,15 @@
     decode_segwit_address,
     encode_segwit_address,
 )
-
```
stylistic nit: we usually keep two blank lines after imports, so should rather add another one than remove one here (see also https://peps.python.org/pep-0008/#blank-lines).

ismaelsadeeq replied:
Thanks :100:

theStack reviewed (commented):
Concept ACK, thanks for following up!
LGTM, left just one small suggestion below.

ismaelsadeeq committed:
refactor: move address_to_scriptpubkey to address.py

The COINBASE_MATURITY constant in blocktools.py is imported in wallet.py.
However, importing address_to_scriptpubkey to blocktools.py will
generate a circular import error. Since the method is related to
addresses, it is best to move it to address.py, which will also
fix the circular import error.

Update imports of address_to_scriptpubkey accordingly.

ismaelsadeeq committed:
refactor: use address_to_scriptpubkey to retrieve addresses scriptpubkey

This commit updates the code by replacing the RPC call used to
decode an address and retrieve its corresponding scriptpubkey
with the address_to_scriptpubkey function. address_to_scriptpubkey
function can now decode all addresses formats, which makes
it more efficient to use.

josibake commented:
re-ACK https://github.com/bitcoin/bitcoin/pull/27349/commits/e47ce42f670fc43859c157766b342509ab5916f9

theStack reviewed (approved):
ACK e47ce42f670fc43859c157766b342509ab5916f9 🌱
